File: Chapter 2-11-NSView compatible UIView/Objective-C/NSView compatible with UIView/NSView+CompatibleUIView.m
//
//  NSView+CompatibleUIView.m
//  MacOBJC
//
//  Created by Fireball on 11/07/15.
//  Copyright (c) 2015 -. All rights reserved.
//

#import "NSView+CompatibleUIView.h"

@implementation NSView (CompatibleUIView)


- (void)setAlpha:(CGFloat)point {
  [self setAlphaValue:point];
}

- (CGFloat)alpha {
  return self.alphaValue;
}

- (void)setCenter:(CGPoint)point {
  CGSize size = self.frame.size;
  
  CGRect newFrame = CGRectZero;
  newFrame.size = size;
  
  CGFloat width = CGRectGetWidth(self.frame);
  CGFloat height = CGRectGetHeight(self.frame);
  
  CGFloat x = floorf(point.x - (width / 2.0f) );
  CGFloat y = floorf(point.y - (height / 2.0f) );
  
  newFrame.origin.x = x;
  newFrame.origin.y = y;
  
  [self setFrame:newFrame];
}

- (CGPoint)center {
  
  return CGPointMake(CGRectGetMidX(self.frame),
                     CGRectGetMidY(self.frame));
  
}

+(void)fadeOut:(NSView*)viewToDissolve
  withDuration:(NSTimeInterval)duration
{
  [NSAnimationContext beginGrouping];
  [[NSAnimationContext currentContext] setDuration:duration];
  [[viewToDissolve animator] setAlphaValue:0.0];
  [NSAnimationContext endGrouping];
  
}

+(void)fadeIn:(NSView*)viewToFadeIn
 withDuration:(NSTimeInterval)duration
{
  [NSAnimationContext beginGrouping];
  [[NSAnimationContext currentContext] setDuration:duration];
  [[viewToFadeIn animator] setAlphaValue:1.0f];
  [NSAnimationContext endGrouping];
}

+ (void)animateWithDuration:(NSTimeInterval)duration
                 animations:(void (^)(void))animations
{
  [NSAnimationContext beginGrouping];
  [[NSAnimationContext currentContext] setDuration:duration];
  animations();
  [NSAnimationContext endGrouping];
}

+ (void)animateWithDuration:(NSTimeInterval)duration
                 animations:(void (^)(void))animations
                 completion:(void (^)(BOOL finished))completion
{
  [NSAnimationContext beginGrouping];
  [[NSAnimationContext currentContext] setDuration:duration];
  animations();
  [NSAnimationContext endGrouping];
  
  if(animations)
  {
    id completionBlock = [completion copy];
    [self performSelector:@selector(runEndBlock:)
               withObject:completionBlock
               afterDelay:duration];
  }
}

+ (void)runEndBlock:(void (^)(void))completionBlock
{
  completionBlock();
}


@end

File: Chapter 2-11-NSView compatible UIView/Swift 2.0/NSView compatible with UIView/NSVIew+CompatibleUIVIew.swift
//
//  NSVIew+CompatibleUIVIew.swift
//  MacSwift
//
//  Created by Fireball on 11/07/15.
//  Copyright (c) 2015 -. All rights reserved.
//

import Foundation
import AppKit

extension NSView {
  
  typealias AnimationHandler  = () -> Void
  typealias CompletionHandler = (finished : Bool) -> Void
  
  
  var center:CGPoint {
    
    get {
      let midX = CGRectGetMidX(frame)
      let midY = CGRectGetMidY(frame)
      
      return CGPointMake(midX, midY)
    }
    
    set(newCenter) {
      var newFrame = CGRectZero
      newFrame.size = frame.size
      
      let myWidth = CGRectGetWidth(frame)
      let myHeight = CGRectGetHeight(frame)
      
      let x = floor(newCenter.x - (myWidth / 2.0) )
      let y = floor(newCenter.y - (myHeight / 2.0) )
      
      newFrame.origin.x = x
      newFrame.origin.y = y
      
      frame = newFrame
    }
    
    
  }
  

  var alpha : CGFloat {
        
        get{
           return alphaValue
        }
        
        set(value) {
           alphaValue = value
        }
        
  }
  
  func fadeInWithDuration (duration : NSTimeInterval) {
          wantsLayer = true
          NSAnimationContext.beginGrouping()
          NSAnimationContext.currentContext().duration = duration
          animator().alphaValue = 1.0
          NSAnimationContext.endGrouping()
  }
  
  func fadeOutWithDuration (duration : NSTimeInterval) {
          wantsLayer = true
          NSAnimationContext.beginGrouping()
          NSAnimationContext.currentContext().duration = duration
          animator().alphaValue = 0.0
          NSAnimationContext.endGrouping()
  }

  func animateWithDuration (duration : NSTimeInterval, animations:AnimationHandler) {
            wantsLayer = true
            NSAnimationContext.runAnimationGroup({ (context: NSAnimationContext) -> Void in
            
            context.duration = duration
            context.allowsImplicitAnimation = true
            
            animations()
            }, completionHandler: { () -> Void in
            
            })
  }
  

  func animateWithDuration (duration : NSTimeInterval, animations:AnimationHandler, completion:CompletionHandler) {
              wantsLayer = true
              
              NSAnimationContext.runAnimationGroup({ (context: NSAnimationContext) -> Void in
              
              context.duration = duration
              context.allowsImplicitAnimation = true
              
              animations()
              }, completionHandler: { () -> Void in
    completion(finished:true)
    
              })
              
  }
}